### ISSUE TITLE
Tracking Issue for `private_macro_use` lint

### ISSUE THREAD

**Issue by bvanjoi** [rust-lang/rust#120192]
This is a tracking issue for the `private_macro_use`, a lint that follows the [breaking change policy](https://github.com/rust-lang/rfcs/blob/master/text/1589-rustc-bug-fix-procedure.md).

### Description

The root of this issue was originally posted in #119301. A macro that should not have been exported was indeed exported. A minimal example to illustrate this issue is shown below:

```rs
// extern_macro.rs
macro_rules! foo { () => {};

// code.rs
#[macro_use]
extern crate extern_macro;

fn main() {
    foo!(); // this should trigger a compile error, but it doesn't
}
```

The problem was triggered by an oversight where a visibility check for the macro, defined in an external crate, was omitted. Now, we have reintroduced this check to prevent such issues from occurring.

### Steps

- [x] Implement: https://github.com/rust-lang/rust/pull/119369
- [ ] Make it as an error

**Comment by tgrushka** [comment]
The new imposition of `private_macro_use` is a really bad idea, and it is going to cause chaos in the Rust community, because `cargo check` on the "offending" crate does not raise this lint, and now it is becoming **deny** and therefore will raise a compile **error** on the downstream crate, and will be **completely invisible** to the upstream author who caused the issue.

This is **utterly ridiculous developer experience** and very anti-Rust. The lint does not provide enough information. It does not help developers to raise compile errors in dependent crates when the upstream crate compiles without error or warning. And if the used macro is truly private, I would think the dependent crate would not compile anyway, so what is the additional use of this lint if it adds absolutely no value beyond what the compile error itself would offer (and indeed would likely provide more useful information)?

It seems to me that `private_macro_use` should be kept **internal to rustc** to keep the rustc developers in check about what private macros they might be inadvertently exporting, and should, if made public, be **totally optional.**

Look at `num-bigint-dig` crate, which is used by `rsa` and others. It will soon no longer compile without a patch because of this lint, especially if it is changed to **deny** by default. I forked and cloned that crate, and could **not** get this lint to show on `cargo check` using any combination of flags on that crate.

This will raise random errors all over the place in any crate that uses even **public** macros such as `vec![]` that do not `pub use alloc::vec`, etc. This is just a major disaster waiting to happen. Please fix it. Take this back to the drawing board and make it:
1. not lint on **any** public macros (clearly it doesn't respect this rule, because `vec![]` is public by definition);
2. lint on the **offending** crate as well as in any downstream crate (currently does **not** lint at all on the offending crate);
3. able to be overridden with something more convenient than `-Aprivate-macro-use` rustc flag or `#![allow(Aprivate_macro_use)]` directive, because this will raise warnings everywhere.

**Comment by tokarevart** [comment]
Just upgraded to newest nightly and also got the warning mentioning `num-bigint-dig`. For now, until this warning works correctly, I've suppressed all future incompat warnings by putting the following in `.cargo/config.toml` in my project
```rust
[future-incompat-report]
frequency = "never"
```

**Comment by xtqqczze** [comment]
> Just upgraded to newest nightly and also got the warning mentioning `num-bigint-dig`.

This has since been fixed in `num-bigint-dig` v0.8.6.
